### PR TITLE
bugfix: bound subtyping depth to avoid reliance on unpredictable stack overflow

### LINKED ISSUE #3057: Segmentation fault: 11 when trying to define a method which returns a new “instance” of a class
I’m sure I brought this up before but I ran into it again and couldn’t find anything about it.

Here’s a fairly minimal example that reproduces the issue when run with `--check`:

```swift
module {
  public class Id<A>(x : A) {
    public func map<B>(f : A -> B) : Id<B> {
      Id(f(x));
    };
  };
};
```

### PR DESCRIPTION
**Pull request by crusso**
The idea is to fail fast and deterministically, especially in VSCode extension. OCaml stack overflow detection is unreliable and sometimes issues OOM, sometimes stack overflow.

Fixes #3057.

- [x] turn failure into compiler error

### CODE REVIEW

File: src/lang_utils/error_codes/M0200.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+Another workaround is to define the problematic method as a separate
+function, outside of the class:
+
+``` motoko
+class Box<T>(v : T) {
+  public let value = v
+};
+
+func map<T, R>(b : Box<T>, f : T -> R) : Box<R> {
+  Box<R>(f (b.value))
```
Comment (crusso) [inline src/lang_utils/error_codes/M0200.md:42]:
```suggestion
  Box<R>(f(b.value))
```

**Review by rvanasa — DISMISSED** [review]

**Review by ggreif — DISMISSED** [review]
LGTM, but can you check whether it catches #3057?

**Comment by crusso** [comment]
> LGTM, but can you check whether it catches #3057?

Yeah, it does. Added test.

File: test/fail/issue-3057.mo [hunk 2]
```diff
@@ -0,0 +1,7 @@
+module {
+  public class Id<A>(x : A) {
+    public func map<B>(f : A -> B) : Id<B> {
+      Id(f(x));
+    };
+  };
+};
```
Comment (crusso) [inline test/fail/issue-3057.mo:7]:
```suggestion
};

```

**Review by ggreif — APPROVED** [review]
LGTM!

**Pull request merged**